rfc/system/5598-deterministic-errors-distributed-training.md

```diff
@@ -0,0 +1,93 @@
+# Deterministic error propagation for distributed (training) tasks
+
+**Authors:**
+
+- @bgedik
+- @eapolinario
+- @fg91
+
+## 1 Executive Summary
+
+Flyte can schedule distributed training jobs leverging e.g. the [kubeflow training operator](https://github.com/kubeflow/training-operator/tree/f55a91d03f23498cdb465ac26c78566228077c51) and its `PyTorchJob`, `TFJob`, `MPIJob`, ...
+
+For these distributed jobs, multiple Kubernetes pods are launched. Any of these worker pods can crash, causing all other worker pods in the distributed job to fail subsequently because one worker disappeared.
+
+Error propagation, in Flyte, happens by the pod entrypoint uploading a file called `error.pb` to blob storage which contains (among other things) the error message and the information whether the error is retriable.
+
+In a distributed training job, all worker pods currently try to create the same `error.pb` file in blob storage - leading to a race condition. It is not guaranteed that the root-cause error is the one being reported to the user and used to determine whether the task can be retried.
```

Comment:
As a side note, the current behavior typically results in the worst outcome, as the latter errors override the former ones, which is the exact opposite of the desired behavior of identifying the first error as the root cause.

Not that PyTorch elastic sets the earliest error as the root cause for the errors that happen within a pod, but across pods we typically end up having the opposite, the latest pod error taking over.

Reply:
Added the first sentence of your remark https://github.com/flyteorg/flyte/pull/5598/commits/68a623a8010a475ecc5306158ade722c4365c2d5

Didn't add the note about elastic because I'd prefer the RFC to not be about one specific plugin and I feel we make the general problem more than apparent.